-1

Tested at: 9489aa9985a52091531ee147c739be05956c8d95

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-22409/26417/summary.html

I found follow errors while testing this PR

**Failed tests:  Build  ClangBuild**

* **Build**:

I found an error when building: 
 
 <pre>Entering library rule at src/RecoLocalTracker/SiPixelClusterizer/test
Entering library rule at src/RecoLocalTracker/SiPixelClusterizer/test
Entering library rule at src/RecoLocalTracker/SiPixelClusterizer/test
Entering library rule at src/RecoLocalTracker/SiPixelClusterizer/plugins
>> Compiling  /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_1_X_2018-03-01-1100/src/Geometry/TrackerGeometryBuilder/test/phase1PixelTopology_t.cpp 
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_1_X_2018-03-01-1100/src/Geometry/TrackerGeometryBuilder/test/phase1PixelTopology_t.cpp:5:75: fatal error: Geometry/TrackerGeometryBuilder/interface/phase1PixelTopology.h: No such file or directory
 #include "Geometry/TrackerGeometryBuilder/interface/phase1PixelTopology.h"
                                                                           ^
compilation terminated.
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_1_X_2018-03-01-1100/src/Geometry/TrackerGeometryBuilder/test/phase1PixelTopology_t.cpp:5:75: fatal error: Geometry/TrackerGeometryBuilder/interface/phase1PixelTopology.h: No such file or directory
 #include "Geometry/TrackerGeometryBuilder/interface/phase1PixelTopology.h"
</pre>

* **Clang**:

I found a compilation error while trying to compile with clang: 
I used this command: 
 scram b vclean && scram build -k -j 16 USER_CXXFLAGS='-fsyntax-only' COMPILER='llvm compile'


 <pre>>> Compiling edm plugin /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_1_X_2018-03-01-1100/src/Geometry/TrackerGeometryBuilder/test/ModuleInfo.cc 
>> Compiling edm plugin /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_1_X_2018-03-01-1100/src/Geometry/TrackerGeometryBuilder/test/TrackerParametersAnalyzer.cc 
>> Compiling edm plugin /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_1_X_2018-03-01-1100/src/Geometry/TrackerGeometryBuilder/test/TrackerDigiGeometryAnalyzer.cc 
>> Compiling edm plugin /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_1_X_2018-03-01-1100/src/Geometry/TrackerGeometryBuilder/test/GeoHierarchyTest.cc 
>> Compiling edm plugin /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_1_X_2018-03-01-1100/src/Geometry/TrackerGeometryBuilder/test/TrackerMapTool.cc 
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_1_X_2018-03-01-1100/src/Geometry/TrackerGeometryBuilder/test/phase1PixelTopology_t.cpp:5:10: fatal error: 'Geometry/TrackerGeometryBuilder/interface/phase1PixelTopology.h' file not found
#include "Geometry/TrackerGeometryBuilder/interface/phase1PixelTopology.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
>> Compiling edm plugin /build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_1_X_2018-03-01-1100/src/Geometry/TrackerGeometryBuilder/plugins/TrackerParametersESModule.cc 
/build/cmsbld/jenkins/workspace/ib-any-integration/CMSSW_10_1_X_2018-03-01-1100/src/Geometry/TrackerGeometryBuilder/test/phase1PixelTopology_t.cpp:5:10: fatal error: 'Geometry/TrackerGeometryBuilder/interface/phase1PixelTopology.h' file not found
</pre>

--- a/CalibTracker/SiPixelESProducers/interface/SiPixelGainCalibrationServiceBase.h
+++ b/CalibTracker/SiPixelESProducers/interface/SiPixelGainCalibrationServiceBase.h
@@ -73,6 +73,8 @@ class SiPixelGainCalibrationServicePayloadGetter : public SiPixelGainCalibration
 
   void    setESObjects(const edm::EventSetup& es ) override;
 
+  thePayloadObject const & payload() const { return *ped; }
+
   std::vector<uint32_t> getDetIds() override;
   double getGainLow() override;
   double getGainHigh() override;

--- a/CondFormats/SiPixelObjects/interface/SiPixelGainCalibrationForHLT.h
+++ b/CondFormats/SiPixelObjects/interface/SiPixelGainCalibrationForHLT.h
@@ -73,6 +73,9 @@ class SiPixelGainCalibrationForHLT {
   float getPedLow() const { return minPed_; }
   float getPedHigh() const { return maxPed_; }
 
+  std::vector<char> const & data() const { return v_pedestals;}
+  std::vector<DetRegistry> const & getIndexes() const { return indexes; }
+
   // Set and get public methods
   void  setData(float ped, float gain, std::vector<char>& vped, bool thisColumnIsDead = false, bool thisColumnIsNoisy = false);
   void  setDeadColumn(const int& nRows, std::vector<char>& vped)  { setData(0, 0 /*dummy values, not used*/, vped, true, false); }
@@ -84,7 +87,7 @@ class SiPixelGainCalibrationForHLT {
   float getPed   (const int& col, const int& row, const Range& range, const int& nCols, bool& isDeadColumn, bool& isNoisyColumn ) const;
   float getGain  (const int& col, const int& row, const Range& range, const int& nCols, bool& isDeadColumn, bool& isNoisyColumn ) const;
 
-  private:
+private:
 
   float   encodeGain(const float& gain);
   float   encodePed (const float& ped);

--- a/DataFormats/GeometrySurface/interface/SOARotation.h
+++ b/DataFormats/GeometrySurface/interface/SOARotation.h
@@ -1,0 +1,158 @@
+#pragma once
+#ifndef DataFormats_GeometrySurface_SOARotation_h
+#define DataFormats_GeometrySurface_SOARotation_h
+
+template <class T> class TkRotation;
+
+// to be moved in an external common library???
+
+/** Rotation matrix used by SOA (as in GPU)
+ */
+
+template <class T>
+class SOARotation {
+public:
+
+  constexpr inline
+  SOARotation(){}
+
+  constexpr inline
+  explicit SOARotation(T) :
+    R11(1), R12(0), R13(0),
+    R21(0), R22(1), R23(0),
+    R31(0), R32(0), R33(1) {}
+
+  constexpr inline
+  SOARotation(T xx, T xy, T xz, T yx, T yy, T yz, T zx, T zy, T zz) :
+    R11(xx), R12(xy), R13(xz),
+    R21(yx), R22(yy), R23(yz),
+    R31(zx), R32(zy), R33(zz) {}
+
+  constexpr inline
+  SOARotation(const T* p) :
+    R11(p[0]), R12(p[1]), R13(p[2]),
+    R21(p[3]), R22(p[4]), R23(p[5]),
+    R31(p[6]), R32(p[7]), R33(p[8]) {}
+
+
+  template <typename U>
+  constexpr inline
+  SOARotation(const TkRotation<U>& a) :
+    R11(a.xx()), R12(a.xy()), R13(a.xz()),
+    R21(a.yx()), R22(a.yy()), R23(a.yz()),
+    R31(a.zx()), R32(a.zy()), R33(a.zz()) {}
+
+  constexpr inline
+  SOARotation transposed() const {
+      return SOARotation(R11, R21, R31,
+			 R12, R22, R32,
+			 R13, R23, R33);
+  }
+
+  // if frame this is to local
+  constexpr inline
+  void multiply(T const vx, T const vy, T const vz,
+                T & ux, T & uy, T & uz) const {
+     ux = R11*vx + R12*vy + R13*vz;
+     uy = R21*vx + R22*vy + R23*vz;
+     uz = R31*vx + R32*vy + R33*vz;
+  }
+
+  // if frame this is to global
+  constexpr inline
+  void multiplyInverse (T const vx, T const vy, T const vz,
+              T & ux, T & uy, T & uz) const {
+     ux = R11*vx + R21*vy + R31*vz;
+     uy = R12*vx + R22*vy + R32*vz;
+     uz = R13*vx + R23*vy + R33*vz;
+  }
+
+
+ // if frame this is to global
+ constexpr inline
+ void multiplyInverse (T const vx, T const vy,
+                       T & ux, T & uy, T & uz) const {
+     ux = R11*vx + R21*vy;
+     uy = R12*vx + R22*vy;
+     uz = R13*vx + R23*vy;
+  }
+
+
+  constexpr inline
+  T const &xx() const { return R11; }
+  constexpr inline
+  T const &xy() const { return R12; }
+  constexpr inline
+  T const &xz() const { return R13; }
+  constexpr inline
+  T const &yx() const { return R21; }
+  constexpr inline
+  T const &yy() const { return R22; }
+  constexpr inline
+  T const &yz() const { return R23; }
+  constexpr inline
+  T const &zx() const { return R31; }
+  constexpr inline
+  T const &zy() const { return R32; }
+  constexpr inline
+  T const &zz() const { return R33; }
+
+private:
+
+  T R11, R12, R13;
+  T R21, R22, R23;
+  T R31, R32, R33;
+};
+
+
+template <class T>
+class SOAFrame {
+public:
+
+  constexpr inline
+  SOAFrame(){}
+
+  constexpr inline
+  SOAFrame(T ix, T iy, T iz, SOARotation<T> const & irot) :
+     px(ix), py(iy), pz(iz), rot(irot){}
+
+  constexpr inline
+  SOARotation<T> const & rotation() const { return rot; }
+
+  constexpr inline
+  void toLocal(T const vx, T const vy,	T const vz,
+               T & ux, T & uy, T & uz) const {
+      rot.multiply(vx-px, vy-py, vz-pz, ux, uy, uz);
+  }
+
+
+  constexpr inline
+  void toGlobal(T const vx, T const vy, T const vz,
+                T & ux, T & uy, T & uz) const {
+       rot.multiplyInverse(vx, vy, vz, ux, uy, uz);
+       ux+=px; uy+=py; uz+=pz;
+  }
+
+  constexpr inline
+  void toGlobal(T const vx, T const vy,
+                T & ux, T & uy, T & uz) const {
+       rot.multiplyInverse(vx, vy, ux, uy, uz);
+       ux+=px; uy+=py; uz+=pz;
+  }
+
+
+  constexpr inline
+  T x() const { return px; }
+  constexpr inline
+  T y()	const { return py; }
+  constexpr inline
+  T z()	const { return pz; }
+
+private:
+
+  T px, py, pz;
+  SOARotation<T> rot;
+
+};
+
+#endif // DataFormats_GeometrySurface_SOARotation_h

--- a/DataFormats/GeometrySurface/test/BuildFile.xml
+++ b/DataFormats/GeometrySurface/test/BuildFile.xml
@@ -1,16 +1,16 @@
-<use   name="DataFormats/GeometrySurface"/>
-<use   name="cppunit"/>
-<use   name="boost"/>
-<bin   file="FrameTransformTest.cpp">
-</bin>
-<bin   file="OldFrameTransformTest.cpp">
-</bin>
+<use name="boost"/>
+<use name="cppunit"/>
+<use name="DataFormats/GeometrySurface"/>
 
-<bin   file="PrecisionMixSurface.cpp">
+<bin file="FrameTransformTest.cpp">
 </bin>
-<bin   file="referencecounted_t.cpp">
+<bin file="OldFrameTransformTest.cpp">
 </bin>
-<bin   file="testGeometricSort.cpp" name="testGeometricSort">
+<bin file="PrecisionMixSurface.cpp">
 </bin>
-<bin   file="Bounds_t.cpp">
+<bin file="referencecounted_t.cpp">
+</bin>
+<bin file="testGeometricSort.cpp" name="testGeometricSort">
+</bin>
+<bin file="Bounds_t.cpp">
 </bin>

--- a/DataFormats/GeometrySurface/test/FrameTransformTest.cpp
+++ b/DataFormats/GeometrySurface/test/FrameTransformTest.cpp
@@ -1,56 +1,52 @@
+#include <cmath>
 #include <iostream>
+
+#include "DataFormats/GeometrySurface/interface/SOARotation.h"
 #include "DataFormats/GeometrySurface/interface/TkRotation.h"
 #include "DataFormats/GeometrySurface/interface/GloballyPositioned.h"
 #include "DataFormats/GeometrySurface/interface/BoundPlane.h"
-//#include "CommonReco/RKPropagators/interface/FrameChanger.h"
 #include "DataFormats/GeometrySurface/interface/Cylinder.h"
 
-#include <cmath>
-
 using namespace std;
+
 template<typename T>
-void  go() {
+void go() {
 
     typedef TkRotation<T>                   Rotation;
+    typedef SOARotation<T>                  SRotation;
     typedef GloballyPositioned<T>           Frame;
-    typedef typename Frame::PositionType             Position;
-    typedef typename Frame::GlobalVector             GlobalVector;
-    typedef typename Frame::GlobalPoint              GlobalPoint;
-    typedef typename Frame::LocalVector              LocalVector;
-    typedef typename Frame::LocalPoint               LocalPoint;
+    typedef SOAFrame<T>                     SFrame;
+    typedef typename Frame::PositionType    Position;
+    typedef typename Frame::GlobalVector    GlobalVector;
+    typedef typename Frame::GlobalPoint     GlobalPoint;
+    typedef typename Frame::LocalVector     LocalVector;
+    typedef typename Frame::LocalPoint      LocalPoint;
 
-    std::cout << "size of Rot     " << sizeof(Rotation) << std::endl;
-    std::cout << "size of Pos     " << sizeof(Position) << std::endl;
-    std::cout << "size of Point   " << sizeof(GlobalPoint) << std::endl;
-    std::cout << "size of Frame   " << sizeof(Frame) << std::endl;
+    std::cout << "size of Rot       " << sizeof(Rotation)    << std::endl;
+    std::cout << "size of Pos       " << sizeof(Position)    << std::endl;
+    std::cout << "size of Point     " << sizeof(GlobalPoint) << std::endl;
+    std::cout << "size of Frame     " << sizeof(Frame)       << std::endl;
+    std::cout << "size of soa Rot   " << sizeof(SRotation)   << std::endl;
+    std::cout << "size of soa Frame " << sizeof(SFrame)      << std::endl;
 
     double a = 0.01;
     double ca = cos(a);
     double sa = sin(a);
 
     Rotation r1(ca, sa, 0,
-		-sa, ca, 0,
-		0,   0,  1);;
-    Frame f1(Position(2,3,4), r1);
+               -sa, ca, 0,
+                 0,  0, 1);;
+    Frame f1(Position(2, 3, 4), r1);
     cout << "f1.position() " << f1.position() << endl;
     cout << "f1.rotation() " << endl << f1.rotation() << endl;
 
-    Rotation r2( GlobalVector( 0, 1 ,0), GlobalVector( 0, 0, 1));
-    Frame f2(Position(5,6,7), r2);
+    Rotation r2( GlobalVector( 0, 1 , 0), GlobalVector( 0, 0, 1));
+    Frame f2(Position(5, 6, 7), r2);
     cout << "f2.position() " << f2.position() << endl;
     cout << "f2.rotation() " << endl << f2.rotation() << endl;
 
-    // transform f2 to f1 so that f1 becomes the "global" frame of f3
-    // Rotation r3 = r2.multiplyInverse(r1);
-    // Rotation r3 = r2*r1;
-
-    // Rotation r3 = r1*r2;
-    // Rotation r3 = r1*r2.transposed();
-    // Rotation r3 = r1.transposed()*r2;
-    // Rotation r3 = r1.transposed()*r2.transposed();
-    // Rotation r3 = r2*r1;
     Rotation r3 = r2*r1.transposed();
-
+    SRotation sr3(r3);
 
     GlobalPoint pos2(f2.position());
     LocalPoint lp3 = f1.toLocal(pos2);
@@ -58,8 +54,20 @@ void  go() {
     cout << "f3.position() " << f3.position() << endl;
     cout << "f3.rotation() " << endl << f3.rotation() << endl;
 
-// test
-    GlobalPoint gp( 11,22,33);
+    SFrame sf1(f1.position().x(),
+               f1.position().y(),
+               f1.position().z(),
+               f1.rotation()
+              );
+
+    SFrame sf3(f3.position().x(),
+               f3.position().y(),
+               f3.position().z(),
+               sr3
+              );
+
+    // test
+    GlobalPoint gp(11, 22, 33);
     LocalPoint p_in1 = f1.toLocal( gp);
     typename Frame::ToLocal ff1(f1);
     LocalPoint p_in2 = f2.toLocal( gp);
@@ -69,25 +77,19 @@ void  go() {
     cout << "p_in2 " << p_in2 << endl;
     cout << "p_in3 " << p_in3 << endl;
 
-    LocalPoint p_in1_from3( f3.toGlobal(p_in3).basicVector());
-    cout << "p_in1_from3 + " << p_in1_from3 << endl;
+    LocalPoint p_in1_from3(f3.toGlobal(p_in3).basicVector());
+    cout << "p_in1_from3 " << p_in1_from3 << endl;
 
-    BoundPlane plane(f2.position(), f2.rotation());
-
-//     FrameChanger<double> changer;
-//     FrameChanger<double>::PlanePtr pp = changer.toFrame( plane, f1);
-
-/*
-    FrameChanger changer;
-    FrameChanger::PlanePtr pp = changer.transformPlane( plane, f1);
-    
-    LocalPoint p_in2p = plane.toLocal( gp);
-    LocalPoint p_in3p = pp->toLocal( GlobalPoint(p_in1.basicVector()));
-    cout << "p_in2p " << p_in2p << endl;
-    cout << "p_in3p " << p_in3p << endl;
-*/
-
-
+    T xx, yy, zz;
+    sf1.toLocal(gp.x(), gp.y(), gp.z(),
+                xx, yy, zz);
+    cout << "p_in1 soa " << xx<<','<<yy<<','<< zz << endl;
+    sf3.toLocal(p_in1.x(), p_in1.y(), p_in1.z(),
+                 xx, yy, zz);
+    cout << "p_in3 soa " << xx<<','<<yy<<','<< zz << endl;
+    sf3.toGlobal(p_in3.x(), p_in3.y(), p_in3.z(),
+                 xx, yy, zz);
+    cout << "p_in1_from3 soa " << xx<<','<<yy<<','<< zz << endl;
 }
 
 void cyl() {
@@ -99,31 +101,31 @@ void cyl() {
   typedef typename Frame::GlobalPoint              GlobalPoint;
   typedef typename Frame::LocalVector              LocalVector;
   typedef typename Frame::LocalPoint               LocalPoint;
-  
+
   // cylinder
   {
     std::cout << " Trivial Cylinder" << std::endl;
     Rotation ll( GlobalVector( 1, 0, 0), GlobalVector( 0, 1, 0));
-    Position p0(0,0,0);
-    Cylinder cyl(5.,p0, ll);
-    Plane t = cyl.fastTangent(GlobalPoint(3.,4.,1.));
+    Position p0(0, 0, 0);
+    Cylinder cyl(5., p0, ll);
+    Plane t = cyl.fastTangent(GlobalPoint(3., 4., 1.));
     std::cout << t.position() << '\n' << t.rotation() << std::endl;
     std::cout << t.rotation().x()*cyl.rotation().z().cross( (t.position()-cyl.position()).basicVector() ).unit() << std::endl;
   }
-  
+
   {
     std::cout << " rotated, displaced Cylinder" << std::endl;
     Rotation ll( Basic3DVector<T>( 1, 1, 1), .3);
-    Cylinder cyl(5.f, Position(2,-1,3), ll);
-    Plane t = cyl.fastTangent(LocalPoint(3.,4.,1.));
+    Cylinder cyl(5.f, Position(2,-1, 3), ll);
+    Plane t = cyl.fastTangent(LocalPoint(3., 4., 1.));
     std::cout << t.position() << '\n' << t.rotation() << std::endl;
     std::cout << t.rotation().x()*cyl.rotation().z().cross( (t.position()-cyl.position()).basicVector() ).unit() << std::endl;
   }
-  
+
 }
 
 int main() {
-  
+
   go<float>();
   cyl();
   std::cout << std::endl;

--- a/RecoLocalTracker/SiPixelClusterizer/test/BuildFile.xml
+++ b/RecoLocalTracker/SiPixelClusterizer/test/BuildFile.xml
@@ -1,35 +1,33 @@
-<use   name="Geometry/TrackerGeometryBuilder"/>
-<use   name="FWCore/Framework"/>
-<use   name="DataFormats/Common"/>
-<use   name="FWCore/ParameterSet"/>
-<use   name="FWCore/PluginManager"/>
-<use   name="clhep"/>
-<use   name="boost"/>
-<use   name="root"/>
-<use   name="DataFormats/L1GlobalTrigger"/>
-<use   name="CondFormats/L1TObjects"/>
-<use   name="L1Trigger/GlobalTriggerAnalyzer"/>
-<use   name="CommonTools/UtilAlgos"/>
-<use   name="Geometry/Records"/>
-<use   name="DataFormats/DetId"/>
-# for tracks
-<use   name="TrackingTools/TrajectoryState"/>
-<use   name="TrackingTools/TrackFitters"/>
-<use   name="TrackingTools/TransientTrack"/>
-<use   name="DataFormats/VertexReco"/>
-# for lumi
-<use   name="DataFormats/Luminosity"/>
-<use   name="RecoLuminosity/LumiProducer"/>
-#
-<flags   EDM_PLUGIN="1"/>
-<library   file="ReadPixClusters.cc" name="ReadPixClusters">
+<use name="boost"/>
+<use name="clhep"/>
+<use name="root"/>
+<use name="CommonTools/UtilAlgos"/>
+<use name="CondFormats/L1TObjects"/>
+<use name="DataFormats/Common"/>
+<use name="DataFormats/DetId"/>
+<use name="DataFormats/L1GlobalTrigger"/>
+<use name="DataFormats/Luminosity"/>
+<use name="DataFormats/VertexReco"/>
+<use name="FWCore/Framework"/>
+<use name="FWCore/ParameterSet"/>
+<use name="FWCore/PluginManager"/>
+<use name="Geometry/Records"/>
+<use name="Geometry/TrackerGeometryBuilder"/>
+<use name="L1Trigger/GlobalTriggerAnalyzer"/>
+<use name="RecoLuminosity/LumiProducer"/>
+<use name="TrackingTools/TrackFitters"/>
+<use name="TrackingTools/TrajectoryState"/>
+<use name="TrackingTools/TransientTrack"/>
+
+<library file="ReadPixClusters.cc" name="ReadPixClusters">
+  <flags EDM_PLUGIN="1"/>
 </library>
-<flags   EDM_PLUGIN="1"/>
-<library   file="TestClusters.cc" name="TestClusters">
+<library file="TestClusters.cc" name="TestClusters">
+  <flags EDM_PLUGIN="1"/>
 </library>
-<flags   EDM_PLUGIN="1"/>
-<library   file="TestWithTracks.cc" name="TestWithTracks">
+<library file="TestWithTracks.cc" name="TestWithTracks">
+  <flags EDM_PLUGIN="1"/>
 </library>
-<flags   EDM_PLUGIN="1"/>
-<library   file="Triplet.cc" name="Triplet">
+<library file="Triplet.cc" name="Triplet">
+  <flags EDM_PLUGIN="1"/>
 </library>

